program-runtime/src/compute_budget.rs

```diff
@@ -102,35 +102,54 @@ impl ComputeBudget {
+
         let error = TransactionError::InstructionError(0, InstructionError::InvalidInstructionData);
-        // Compute budget instruction must be in the 1st 3 instructions (avoid
-        // nonce marker), otherwise ignored
-        for (program_id, instruction) in message.program_instructions_iter().take(3) {
+        for (i, (program_id, instruction)) in message.program_instructions_iter().enumerate() {
             if compute_budget::check_id(program_id) {
-                match try_from_slice_unchecked(&instruction.data) {
-                    Ok(ComputeBudgetInstruction::RequestUnits {
-                        units,
-                        additional_fee,
-                    }) => {
-                        self.max_units = units.min(MAX_UNITS) as u64;
-                        requested_additional_fee = additional_fee as u64;
-                    }
-                    Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
-                        if !requestable_heap_size
-                            || bytes > MAX_HEAP_FRAME_BYTES
-                            || bytes < MIN_HEAP_FRAME_BYTES as u32
-                            || bytes % 1024 != 0
-                        {
-                            return Err(error);
+                // don't include request instructions in default max calc
+                num_instructions = num_instructions.saturating_sub(1);
+
+                // Compute budget instruction must be in the 1st 3 instructions (avoid
+                // nonce marker), otherwise ignored
+                if i < 3 {
+                    match try_from_slice_unchecked(&instruction.data) {
+                        Ok(ComputeBudgetInstruction::RequestUnits {
+                            units,
+                            additional_fee,
+                        }) => {
+                            requested_units = Some(units as u64);
+                            requested_additional_fee = additional_fee as u64;
+                        }
+                        Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
+                            if !requestable_heap_size
+                                || bytes > MAX_HEAP_FRAME_BYTES
+                                || bytes < MIN_HEAP_FRAME_BYTES as u32
+                                || bytes % 1024 != 0
+                            {
+                                return Err(error);
+                            }
+                            self.heap_size = Some(bytes as usize);
                         }
-                        self.heap_size = Some(bytes as usize);
+                        _ => return Err(error),
                     }
-                    _ => return Err(error),
                 }
             }
         }
+
+        self.max_units = if default_units_per_instruction {
+            requested_units
+                .or_else(|| Some(num_instructions.saturating_mul(DEFAULT_UNITS as usize) as u64))
```

Comment:
I thought we were just requesting `DEFAULT_UNITS` for the entire TX if unspecified, not retaining today's behavior.  I guess this is OK, but I'd rather put some pressure on the offenders to change their ways

Reply:
This is the smoothest transition to tx-wide caps without breaking folks and the additional fee will schedule folks appropriately.  I'm still a fan of charging a requested CU based fee to encourage low CU requests.  @aeyakovenko using for additional fee based motivation, open market.

Reply:
I think we should still cap the number of instructions for this calculation.. Like maybe 10 instructions? If you're doing more than that, you're a bot and can add the compute budget ix yourself.

Reply:
It's capped at a total maximum unit which should be sufficient, what does capping the ix's doing for us?  Seems artificial to say that if there are 10 instructions it must be a bot

Reply:
> It's capped at a total maximum unit which should be sufficient

Ah ok, this works too! I clearly didn't read the PR 😅

Reply:
> This is the smoothest transition to tx-wide caps without breaking folks

Yep. I was just fully prepared to break the right folks :wink: